Extension/src/Automation/Async/signal.ts

```diff
@@ -0,0 +1,92 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import { ManualPromise } from './manual-promise';
+import { Resolveable } from './resolvable';
+
+/**
+ * A signal is an externally fulfillable promise ( @see ManualPromise )
+ * that once awaited, will automatically reset to the unwaited state
```

Comment:
>awaited [](http://example.com/codeflow?start=13&length=7)

semantics: I think it would be more correct to say that once it is fulfilled it will be reset automatically.  The await itself does not reset this promise (i.e. it could be awaited in more than one place before the promise is fulfilled), rather the resolve/reject resets the promise.

Reply:
Good Point